### PR TITLE
Add a test of writing a signedbyte dataset

### PR DESCRIPTION
**Pull request by sgillies**
See #3984. To try to reproduce https://github.com/mapbox/rasterio/issues/2180.

### CODE REVIEW

File: autotest/gcore/numpy_rw.py [hunk 1]
```diff
@@ -218,6 +218,22 @@ def test_numpy_rw_10():
...
+    ds.GetRasterBand(1).WriteArray(ar)
+    ds.GetRasterBand(1).SetNoDataValue(-1)
+    ds = None
+
+    ds = gdal.Open('/vsimem/signed8.tif')
+    ar2 = ds.ReadAsArray()
+    ds = None
+    gdal.Unlink('/vsimem/signed8.tif')
+
+    assert ar2[0][0] == -1 and ar2[0][1] == -1
```
Comment (sgillies) [inline autotest/gcore/numpy_rw.py:235]: @rouault the key difference from this to `test_numpy_rw_10` is that we set a nodata value and all array cells written to the dataset have that value (in int8).

Comment (sgillies) [reply autotest/gcore/numpy_rw.py:235]: I've checked that `ar = numpy.array([[-1, -1]], dtype=numpy.int8).astype(numpy.uint8)` does what GDAL requires. All the items of `ar` are `255`.

Comment (sgillies) [reply autotest/gcore/numpy_rw.py:235]: I think I'm able to pin it down to a nodata issue. If we write an array that has all nodata values, there is an error. See https://github.com/mapbox/rasterio/commit/fb065a0f8e437156ccd101c01dd533e96d8c5139#diff-53e13383b76f631d209604d01e8dd23b11ebf681f3a5c930f1a9b38f507bcc44R8.

**Comment by rouault** [comment]
Reproducer incorporated in fix of https://github.com/OSGeo/gdal/pull/4001

**Pull request closed without merging**